### ISSUE TITLE
Preflight support for way better cache hit rate / utilization

### ISSUE THREAD

**Issue by dkarlovi** [darkweak/souin#277]
This concept is a bit complex, so I'll link to the place where they go in detail how it works:
https://foshttpcache.readthedocs.io/en/stable/user-context.html

To explain it here still, imagine this use case:
1. user `USER1` logged in to a e-commerce site on desktop, session ID `sess_desktop_1`
2. user also logged in on mobile, session ID `sess_mobile_1`
3. we implement ESI, many of those responses can also be cached and *shared* between those two sessions
4. the premise is to generate a "preflight request" which will modify the cache keys: in our case, it will remove session ID and add user ID
5. cache makes the preflight request which explains how to modify the request for it to be more cacheable (this request is also served from cache)
6. cache modifies the request before looking it up
7. this means both mobile and desktop sessions can reuse the same cached ESI responses even though they don't share the same cache keys by default

The value in the example is limited, but we can have situations where we have many users / clients in a cohort and we can reuse the same cache entry for all of them. For example: news portal, guests and any regular users see articles rendered for the public, any users with `role=editor` see a different version with editing tooling embedded. This implemented in public facing or API situations can mean a large difference in performance.

**Comment by darkweak** [comment]
TLDR;
Not possible, because the authenticated requests cannot be cached. An HTTP cache is a shared one not a private one.

**Comment by dkarlovi** [comment]
I think the DR part is important here, this is a viable approach used with Varnish which is also a HTTP cache. Without something like this, Souin will never be able to attain a hit rate Varnish is.

The cache is shared, just not by the entire userbase but by smaller cohorts.

**Comment by dkarlovi** [comment]
> the authenticated requests cannot be cached

I just noticed you said this as a fact. Does that mean Souin just blindly passes any requests with a cookie / auth header?

Because those requests should be cached and, again, you're able to do it with Varnish. Having your whole caching layer disable itself just by the user logging in invalidates the whole stack immediately because you can't rely on it to bear the brunt of the traffic.

**Comment by dkarlovi** [comment]
To expand on this, here's the recipe the linked library ships for Varnish which normalizes the cookie header before making the preflight request to establish the context:
https://foshttpcache.readthedocs.io/en/stable/varnish-configuration.html#cleaning-the-cookie-header

**Comment by darkweak** [comment]
The VCL is a way to add your business logic in varnish and that's not possible in Souin.  
You cannot share the cache between `sess_desktop_1` and `sess_mobile_1` because atm you cannot add your business logic.

If you send an Authorization header, Souin will just pass the request to the upstream, won't use a cached response nor store the upstream response.

**Comment by dkarlovi** [comment]
OK, I understand, that's fair. :+1: 

But how does that reflect with for example [API Platform integration](https://github.com/api-platform/core/pull/4540)? API Platform relies on Varnish being used/setup basically as described in the linked library's docs, and API requests are commonly protected by an Authorization header and you still want to cache it in many cases.

It also means the work we added to Go-ESI (passing headers to child requests) doesn't make sense since the request which would have been satisfied from the cache (with ESI tags embedded in the response, so it's safe) will just be passed to the app too.

It also means basic things like serving static files will always hit the origin if the request has a cookie set. :thinking: 

I'm not sure I'd consider this "business logic", it's pretty boilerplate (meaning, everyone using Varnish is using the same logic), which is why the linked library can ship it.

OK closing and switching to Varnish, I guess. :)

**Comment by darkweak** [comment]
The specification is clear, the cache MAY store the authenticated response so it's not mandatory. It wasn't on the roadmap but now there is a PR for that. 
When you store authenticated response you're storing sensitive data, and you must integrate the private cache logic! Anyway I already answered there is an opened PR for that https://github.com/darkweak/souin/pull/283

**Comment by dkarlovi** [comment]
@darkweak while I understand you have a firm stance here and I respect your POV, a modern & robust caching solution must provide a mechanism for a developer to be able to handle this case.

For example, imagine this scenario:

1. I'm logged into the news portal as a regular user
2. I request the homepage
3. the backend creates a homepage with ESI blocks, but otherwise it should be shared between **all** users, that response has `Cache-Control: public` since it's perfectly safe to store it as-is (the private parts are ESI'd)
4. my next request for the homepage will serve that public cache item, the ESI parts are resolved (and can in the future also be cached), the cache system can't just decide to pass any request as soon as there's cookies, that's SUPER unreliable for the vendor since the system is basically circumvented by just setting a single HTTP header, I can easily DDoS your site from my laptop

**My proposal is not to provide a VCL equivalent** but more incorporate exactly what the linked library does via VCL as few configurable procedures, nothing more.

**Comment by darkweak** [comment]
That's why I said I opened the following PR https://github.com/darkweak/souin/pull/283. When this one will be merged, any authenticated requests can be stored and only served to the authenticated users. With that Souin will be able to work as a private cache.
